File: 290-word-pattern.java
class Solution {
    public boolean wordPattern(String pattern, String s) {
        String[] words = s.split(" ");
        if(pattern.length() != words.length) { return false; }
        
        Map<Character, Integer> letterIndexes = new HashMap<>();
        Map<String, Integer> wordIndexes = new HashMap<>();
        for(int i = 0; i < words.length; i++) {
            char letter = pattern.charAt(i);
            String word = words[i];
            if(!letterIndexes.getOrDefault(letter, -1).equals(wordIndexes.getOrDefault(word, -1))) {
                    return false;
            }
            
            if(!letterIndexes.containsKey(letter)) {
                letterIndexes.put(letter, i);
                wordIndexes.put(word, i);
            }
            
        }
        
        return true;
    }
}

File: top150/290_word_pattern.py
    words = s.split()
    if len(pattern) != len(words):
        return False
    
    letterFirstIdxMap = {}
    wordFirstIdxMap = {}
    for i in range(len(pattern)):
        print(i,letterFirstIdxMap,wordFirstIdxMap)
        if (pattern[i] in letterFirstIdxMap and words[i] not in wordFirstIdxMap) or (pattern[i] not in 
        letterFirstIdxMap and words[i] in wordFirstIdxMap):
            return False
        elif pattern[i] in letterFirstIdxMap and words[i] in wordFirstIdxMap:
            if letterFirstIdxMap[pattern[i]] != wordFirstIdxMap[words[i]]:
                return False
        else:
            letterFirstIdxMap[pattern[i]] = i
            wordFirstIdxMap[words[i]] = i
            
    return True
